open-api/rest-catalog-open-api.yaml

```diff
@@ -2838,6 +2988,69 @@ components:
           additionalProperties:
             type: string
 
+    PreplanTableRequest:
+      type: object
+      required:
+        - select
+        - filter
+      properties:
+        select:
+          description: A list of the selected column names. If referencing a column with a nested field, ensure full name is joined by `.` such as `a.b`
+          type: array
+          items:
+            type: string
+        filter:
+          $ref: '#/components/schemas/Expression'
+        case-sensitive:
+          description: Indicates whether column selection and filtering should be case sensitive
+          type: boolean
+          default: true
+        snapshot-id:
+          description: Defaults to the table's current snapshot; This option is not allowed when `snapshot-range` is present in the request.
+          type: integer
+          format: int64
+        snapshot-range:
+          description: A JSON list containing exactly 2 snapshot IDs representing the start (exclusive) and end (inclusive) snapshots. This option is not allowed when `snapshot-id` is present in the request.
+          type: array
+          items:
+            type: integer
+            format: int64
+
+    PlanTableRequest:
+      type: object
+      required:
+        - select
+      properties:
+        select:
+          description: A list of the selected column names
+          type: array
+          items:
+            type: string
+        filter:
+          $ref: '#/components/schemas/Expression'
+        case-sensitive:
+          description: Indicates whether column selection and filtering should be case sensitive
+          type: boolean
+          default: true
+        stats-fields:
+          description: A list of field names for which stats should be included
+          type: array
+          items:
+            type: string
+        plan-task:
+          $ref: '#/components/schemas/PlanTask'
+        snapshot-id:
```

Comment:
I think we should also include a `ref` (tag/branch) and `as-of` timestamp so that the client can make use of all the scan options without requiring the client to parse the table metadata to identify the appropriate snapshot id (that should be handled by the catalog)

Reply:
This should probably be modeled as: `oneOf`: ref, snapshot, snapshot-range, as-of

Reply:
@danielcweeks, this was my suggestion because I think it makes the API simpler. Rather than having several different options and having a more complicated spec, we can collapse all of those into the snapshot the client actually resolves.

That also avoids cases where the client ends up reading a different snapshot than it thinks is current.

Reply:
Here's the context: https://github.com/apache/iceberg/pull/9695#discussion_r1494914340